### ISSUE TITLE
srb2 3ds 2.2.11 ( maybe 2.2.13)

### ISSUE THREAD

**Issue by Lazercode21** [derrekr/srb2_3ds#19]
Could you update the port to use features like multiplayer with the pc game, special stages, mods and 3d models.

**Comment by derrekr** [comment]
Hi! mods are already supported via [autoexec](https://wiki.srb2.org/wiki/Console_script#autoexec.cfg).
I am currently preparing to re-sync the srb2 source code together with @wallefan (which will be an experiment, idk of the 3ds can handle v2.2)

**Comment by bitten2up** [comment]
> Hi! mods are already supported via [autoexec](https://wiki.srb2.org/wiki/Console_script#autoexec.cfg). I am currently preparing to re-sync the srb2 source code together with @wallefan (which will be an experiment, idk of the 3ds can handle v2.2)

glad to see this being worked on again, 2.2 is possible, i currently have the software renderer running the special stages, would be willing to help out on this
what i have been working on is porting egl to the 3ds so then we can use android's gles renderer

**Comment by derrekr** [comment]
> glad to see this being worked on again, 2.2 is possible, i currently have the software renderer running the special stages, would be willing to help out on this what i have been working on is porting egl to the 3ds so then we can use android's gles renderer

interesting! <s>When you say "2.2 is possible", what's the performance with the sw renderer currently?</s>

Porting egl seems like a lot of work, and ...I don't think it is necessary?
We wanted to build on srb2_3ds' existing 3ds renderer backend (see https://github.com/derrekr/srb2_3ds/blob/master/source/nds/i_video.c and https://github.com/derrekr/srb2_3ds/blob/master/source/nds/r_nds3d.c ). It's based on citro3d, which uses the 3ds' gpu (so it is basically an opengl-like hw renderer). It's pretty optimized at this point, I think.
Some changes will be required for the new 2.2 renderer interface, but the main part should be the same(?)

Anyway, we haven't actually started working on 2.2 integration yet.
I think we could use your experience there.

**Comment by bitten2up** [comment]
> interesting! When you say "2.2 is possible", what's the performance with the sw renderer currently?
> 
> Porting egl seems like a lot of work, and ...I don't think it is necessary? We wanted to build on srb2_3ds' existing 3ds renderer backend (see https://github.com/derrekr/srb2_3ds/blob/master/source/nds/i_video.c and https://github.com/derrekr/srb2_3ds/blob/master/source/nds/r_nds3d.c ). It's based on citro3d, which uses the 3ds' gpu (so it is basically an opengl-like hw renderer). It's pretty optimized at this point, I think. Some changes will be required for the new 2.2 renderer interface, but the main part should be the same(?)
> 
> Anyway, we haven't actually started working on 2.2 integration yet. I think we could use your experience there.

yeah that's fine, when i was trying to do it, i realized that it would be difficult to just merge off of it, [later i tried to change the hardware renderer to use citro3d](https://git.do.srb2.org/bitten2up/SRB2/-/blob/3ds-real/src/hardware/r_ctr/r_ctr.c) which didn't really go well  
  
the main reason I didn't go with using it (or just porting the older 3ds renderer) was because im not too familiar with citro3d though i am willing to learn

the egl thing was because i had gles but sdl needed egl as a wrapper

edit: i thought i did this, this should be where i tried adapting the 3ds port's render code, https://github.com/bitten2up/SRB2/blob/3ds-real/src/dummy/

**Comment by bitten2up** [comment]
well it runs ig...
not well though


![1707090209](https://github.com/derrekr/srb2_3ds/assets/55027142/1221d578-ad4f-457d-ac65-da66bb07515d)

**Comment by Finlito** [comment]
sorry, are you still working on porting 2.2?

**Comment by bitten2up** [comment]
> sorry, are you still working on porting 2.2?

actually yes, currently im working on getting it to boot further, but it can get to the intro... *then crash on the first slide*

![image](https://github.com/user-attachments/assets/040d3c15-d46d-4995-abb4-aac609e6c18a)
the sdl verson can get farther, but is unbeatable

**Comment by soniola** [comment]
could you make srb2 2.1.22 running on a old 3ds?

**Comment by bitten2up** [comment]
> could you make srb2 2.1.22 running on a old 3ds?  

im just going to post this  

![image](https://github.com/user-attachments/assets/c9cc29d2-48a0-4cba-98df-db170ef3f5b6)

**Comment by soniola** [comment]
that means its not impossible, but the thing is, reducing the render distance and getting half of the 3ds resolution can help but idk where those files are in the source code

**Comment by soniola** [comment]
on my 3ds doesnt starts, it crash every try

**Comment by derrekr** [comment]
i've tried getting it to run on o3ds with a couple hacks for better perf, but it's still too laggy... as said above, literally every level except the first one is so laggy, it's not playable

**Comment by Axonic3D** [comment]
Forgive me if I'm interrupting, but seeing as 2.2 isn't happening, I'm still curious as to why there isn't at least an update to **2.1.21 at least**, as that's the version which adds PK3 support and improves on the gamepad controls. At least then the game could control better and the best add-ons for 2.1 can be played on the 3DS (granted, performance will vary especially for level mods).

**Comment by bitten2up** [comment]
> seeing as 2.2 isn't happening  

Well 2.2 is happening, currently i have to fix up android because 2.2.14 pre1 decided to drop.  

while 2.2.13 is defiantly possible (and I am working on it), 2.2.14 might be a bit harder because of palette rendering rewriting alot of stuff, but we'll see

**Comment by Axonic3D** [comment]
Bad of me to assume, admittedly. I merely thought so because of each news I see regarding the 2.2 port seem to be that the game runs poorly or not at all. Then again, I don't have many sources for updates asides from here.

What's the game's status on New 3DS? And on the topic of news, is there any particular place where you are sharing them?

**Comment by bitten2up** [comment]
> Bad of me to assume, admittedly. I merely thought so because of each news I see regarding the 2.2 port seem to be that the game runs poorly or not at all. Then again, I don't have many sources for updates asides from here.
> 
> What's the game's status on New 3DS? And on the topic of news, is there any particular place where you are sharing them?

ah dw, if you use software and sdl, gfz1 runs pretty well. other levels on the otherhand aren't so lucky, so i am porting over the renderer used here

**Comment by FreindlyNeighborhoodBeetle** [comment]
how far are you into porting over the renderer? dont want to be annoying but i feel like it been long enough of a timegap to ask

**Comment by bitten2up** [comment]
the intro runs ig... title screen and levels crash currently 

![PXL_20241104_225543534](https://github.com/user-attachments/assets/d6ec902c-184e-4818-ab63-66419d337a03)

**Comment by FreindlyNeighborhoodBeetle** [comment]
well, it's better than the last update where it crashed immediately, so thats good

**Comment by monadoenchant** [comment]
after seeing the witchcraft that was the recent 32x port, I now am convinced anything is possible. really excited for any updates here!

**Comment by derrekr** [comment]
> the recent 32x port

Damn, that is impressive, but apparently it's a total rewrite, so not really comparable

**Comment by HPR69** [comment]
any updates on the update?
